src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -229,7 +232,26 @@ private void initHttpAndDiskCache(
               remoteOptions,
               creds,
               Preconditions.checkNotNull(env.getWorkingDirectory(), "workingDirectory"),
-              digestUtil);
+              digestUtil,
+              new RemoteRetrier(
+                 remoteOptions,
+                 (e) -> {
+                   boolean retry = false;
+                   if (e instanceof ClosedChannelException) {
+                     retry = true;
+                   } else if (e instanceof HttpException) {
+                     retry = true;
+                   } else if (e instanceof IOException) {
+                     String msg = e.getMessage().toLowerCase();
+                     if (msg.contains("connection reset by peer")) {
+                       retry = true;
+                     }
+                   }
```

Comment:
This list of exceptions is based on what we observed on CI, see https://github.com/digital-asset/daml/pull/11238 and https://github.com/digital-asset/daml/pull/11445. It may be that further rules should be added.

As far as I understand, `ClosedChannelException` can also occur in case of build failure when Bazel aborts the build. Is there a way to avoid retry in this case and only retry on legitimate network errors?

Reply:
What kind of HttpException do you expect to catch and retry?

E.g. if server is overloaded and answer with 503 Service Unavailable, then it could make sense to do the opposite, back off and not retry, at least until what is specified by Retry-After header.

Reply:
Sorry for the late reply.

An example that we [saw in our logs](https://github.com/digital-asset/daml/pull/11238#issue-1025314540) was
```
RETRYING: com.google.devtools.build.lib.remote.http.HttpException: 502 Bad Gateway
```

> E.g. if server is overloaded and answer with 503 Service Unavailable, then it could make sense to do the opposite, back off and not retry, at least until what is specified by Retry-After header.

AFAIK, `RemoteRetrier` currently uses an exponential backoff. So, it does already back-off. It's true that it wouldn't respect Retry-After. I'm not too familiar with the internals of `RemoteRetrier`. Does it support that kind of use-case?

Reply:
Just yesterday I encountered a Bazel `HttpException` on a local build with this patch enabled:
```
RETRYING: com.google.devtools.build.lib.remote.http.HttpException: 503 Service Unavailable
Service Unavailable
```
The retry succeeded and the build continued from there on.

Reply:
Maybe the retries should be limited to a specific number of status codes in case of an `HttpException`?

Does it even make sense to retry for *any* client error as resending the request will probably not change the response?! (e.g. for 400, 401, 403 or 405)

Otherwise, retrying on 503 Service Unavailable with the back-off mechanism of the `Retrier` seems reasonable to me, what do you think @ulrfa?

Reply:
Good point, I've added logic to only retry on certain HTTP errors.